### PR TITLE
Upgradeable Contract section

### PR DESCRIPTION
**Pull request by dubzn**
# Description

- 12.6.2 Upgradeable contract
  - General description
  - Replacing contract class hash
  - Upgradeable component (OZ example)
  

- - -
This change is [Reviewable](https://reviewable.io/reviews/cairo-book/cairo-book/629)

### CODE REVIEW

File: src/SUMMARY.md [hunk 1]
```diff
@@ -116,6 +116,8 @@
 
 - [Building Starknet Smart Contracts](./ch13-00-building-starknet-smart-contracts.md)
 
+    - [Deploying and Interacting with a Voting contract](./ch99-01-04-01-voting-contract.md)
```
Comment (dubzn) [inline src/SUMMARY.md:119]: Should I remove this? It seems to have changed since I forked it a few days ago.

**Comment by enitrat** [comment]
hey, yes you can move it to the building advanced contracts page

File: src/SUMMARY.md (hunk 1 above)
Comment (enitrat) [reply src/SUMMARY.md:119]: yes

File: src/ch99-01-04-01-upgradeable-contract.md [hunk 2]
```diff
@@ -0,0 +1,39 @@
+# Upgradeable Contract
+
+Various blockchain networks have devised several approaches to enable contract upgradability, with the proxy patterns being among the most commonly embraced solutions.
```
Comment (enitrat) [inline src/ch99-01-04-01-upgradeable-contract.md:3]: that reads a lot like fluffy chatgpt output. there's nothing wrong in using AI tools but I don't like the gpt style at all - my best advice would be to write the text yourself as best as you can and then pass it to https://www.deepl.com/write/write-mobile if you want it to be more professional and accurate

File: src/ch99-01-04-01-upgradeable-contract.md [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+**OpenZeppelin Contracts** for Cairo provides **`Upgradeable`** to add upgradeability support to your contracts.
+
+Check [OpenZeppelin docs API reference](https://docs.openzeppelin.com/contracts-cairo/0.9.0/api/upgrades). 
+
+```rust,noplayground
+{{#include ../listings/ch99-starknet-smart-contracts/listing_99_12_upgradeable_contract_example/src/lib.cairo}}
+```
+<span class="caption">Upgradeable OZ component usage </span>
+
+## Proxies in Starknet
```
Comment (enitrat) [inline src/ch99-01-04-01-upgradeable-contract.md:35]: i don't understand this part. since proxies don't exist I don t think we should mention it

**Comment by enitrat** [comment]
@dubzn is this ready for review or still draft?

**Comment by dubzn** [comment]
> @dubzn is this ready for review or still draft?

Hi Eni, sorry for the delay, it is ready for review. Thanks for the feedback

**Review by enitrat — CHANGES_REQUESTED** [review]
> Reviewed 1 of 4 files at r1, 2 of 2 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cairo-book/cairo-book/629)* status: 3 of 4 files reviewed, 5 unresolved discussions (waiting on @dubzn)

___
*[`src/ch99-01-04-01-upgradeable-contract.md` line 3 at r3](https://reviewable.io/reviews/cairo-book/cairo-book/629#-Ns_3ZZM2Cqqx8Z75hol:-Ns_3ZZM2Cqqx8Z75hom:b45h) ([raw file](https://github.com/cairo-book/cairo-book/blob/e80ca8bc394c6af30741862c58e1ee7cf78653e4/src/ch99-01-04-01-upgradeable-contract.md#L3)):*
> ```Markdown
> # Upgradeable Contract
> 
> The concept of upgradeable contracts is **natively supported** by Starknet through the [system call](https://docs.starknet.io/documentation/architecture_and_concepts/Smart_Contracts/system-calls-cairo1/) `replace_class`, which gives us the possibility to upgrade interfaces, storage variables and implementations of our contract in a dynamic and flexible way. 
> ```





_Suggestion:_
```Markdown
the class
```
___
*[`src/ch99-01-04-01-upgradeable-contract.md` line 4 at r3](https://reviewable.io/reviews/cairo-book/cairo-book/629#-Ns_4XlMCz0QOBnzRydH:-Ns_4XlMCz0QOBnzRydI:b-ajfx6h) ([raw file](https://github.com/cairo-book/cairo-book/blob/e80ca8bc394c6af30741862c58e1ee7cf78653e4/src/ch99-01-04-01-upgradeable-contract.md#L4)):*
> ```Markdown
> 
> The concept of upgradeable contracts is **natively supported** by Starknet through the [system call](https://docs.starknet.io/documentation/architecture_and_concepts/Smart_Contracts/system-calls-cairo1/) `replace_class`, which gives us the possibility to upgrade interfaces, storage variables and implementations of our contract in a dynamic and flexible way. 
> ```

The class / contract distinction should be mentioned straight away

Starknet separates contracts into classes and instances, making it simple to upgrade a contract's logic without affecting its state.

A _contract class_ is the definition of a contract. It includes Cairo bytecode, entry point names, storage variables addresses, and everything that defines its semantics. Each class is uniquely identified by its _class hash_. A class does not have a storage of its own: it's only a definition of logic.

A _contract instance_ is a deployed contract that corresponds to a class.  They have their own storage and can be called by transactions or other contracts.
___
*[`src/ch99-01-04-01-upgradeable-contract.md` line 5 at r3](https://reviewable.io/reviews/cairo-book/cairo-book/629#-Ns_42wP7zdbq_7XeCGQ:-Ns_42wP7zdbq_7XeCGR:undefined) ([raw file](https://github.com/cairo-book/cairo-book/blob/e80ca8bc394c6af30741862c58e1ee7cf78653e4/src/ch99-01-04-01-upgradeable-contract.md#L5)):*
> ```Markdown
> The concept of upgradeable contracts is **natively supported** by Starknet through the [system call](https://docs.starknet.io/documentation/architecture_and_concepts/Smart_Contracts/system-calls-cairo1/) `replace_class`, which gives us the possibility to upgrade interfaces, storage variables and implementations of our contract in a dynamic and flexible way. 
> 
> In order to upgrade our contract we can expose a function as shown below, note that if you do it this way, anyone will be able to access it.
> ```





_Suggestion:_
```Markdown
In order to upgrade our contract, we need to expose a function that calls `replace_class_syscall` to change the class associated to our contract instance.
```
___
*[`src/SUMMARY.md` line 119 at r3](https://reviewable.io/reviews/cairo-book/cairo-book/629#-Ns_6q501Mi3jrR6QxnS:-Ns_6q501Mi3jrR6QxnT:b-dr6ea8) ([raw file](https://github.com/cairo-book/cairo-book/blob/e80ca8bc394c6af30741862c58e1ee7cf78653e4/src/SUMMARY.md#L119)):*
> ```Markdown
> - [Building Starknet Smart Contracts](./ch13-00-building-starknet-smart-contracts.md)
> 
>   - [Upgradeable contract](./ch99-01-04-01-upgradeable-contract.md)
> ```

Needs to be updated (I think this one can be removed and the other one kept)



_Code quote:_
```Markdown
 - [Upgradeable contract](./ch99-01-04-01-upgradeable-contract.md)
```

**Comment by TAdev0** [comment]
The file should be named ch15... instead of ch99

**Pull request closed without merging**